#!/usr/bin/env python

PACKAGE = "limu_tof_ros"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

lens_type_enum = gen.enum([ gen.const("WideField", int_t, 0, "WideField"),
                             gen.const("StandardField",    int_t, 1, "StandardField"),
                             gen.const("NarrowField", int_t, 2, "NarrowField")],
                             "An enum to set lens type")
                             
frequence_modulation_enum = gen.enum([ gen.const("24MHz", int_t, 0, "24MHz"),
                                       gen.const("12MHz", int_t, 1, "12MHz"),
                                       gen.const("6MHz", int_t, 2, "6MHz"),
                                       gen.const("3MHz", int_t, 3, "3MHz"),
                                       gen.const("1_5MHz", int_t, 4, "1_5MHz"),
                                       gen.const("0_75MHz", int_t, 5, "0_75MHz")],
                                       "An enum to set modulation frequency")

channel_enum = gen.enum([ gen.const("0", int_t, 0, "0"),
                          gen.const("1", int_t, 1, "1"),
                          gen.const("2", int_t, 2, "2"),
                          gen.const("3", int_t, 3, "3"),
                          gen.const("4", int_t, 4, "4"),
                          gen.const("5", int_t, 5, "5"),
                          gen.const("6", int_t, 6, "6"),
                          gen.const("7", int_t, 7, "7"),
                          gen.const("8", int_t, 8, "8"),
                          gen.const("9", int_t, 9, "9"),
                          gen.const("10", int_t, 10, "10"),
                          gen.const("11", int_t, 11, "11"),
                          gen.const("12", int_t, 12, "12"),
                          gen.const("13", int_t, 13, "13"),
                          gen.const("14", int_t, 14, "14"),
                          gen.const("15", int_t, 15, "15")],
                          "An enum to set channel index")

image_type_enum = gen.enum([ gen.const("Distance",    int_t, 1, "Distance"),
                             gen.const("DistranceAmplitude", int_t, 2, "Distance&Amplitude")
                             ],
                             "An enum to set image type")

hdr_mode_enum  =  gen.enum([ gen.const("HDR_off",      int_t, 0, "HDR off"),
                             gen.const("HDR_spatial",  int_t, 1, "HDR spatial"),
                             gen.const("HDR_temporal", int_t, 2, "HDR temporal")],
                             "An enum to set HDR mode")

######  Name                    Type   Reconfiguration level  Description     Default Min  Max
### Acquisition Control
gen.add("lens_type",              int_t,  0,  "Lens type", 1, 0, 2, edit_method=lens_type_enum)
gen.add("frequency_modulation",   int_t,  0,  "Frequency modulation",                  2,  0, 5, edit_method=frequence_modulation_enum)
gen.add("channel",                int_t,  0,  "Chanel",     0, 0, 15, edit_method=channel_enum)
gen.add("image_type",             int_t,  0,  "Image type", 1, 1,  2, edit_method=image_type_enum)
gen.add("hdr_mode",               int_t,  0,  "HDR mode",   2, 0,  2, edit_method=hdr_mode_enum)
gen.add("integration_time_tof_1", int_t,  0,  "Integration time TOF1 [uS]",         50, 0, 4000)
gen.add("integration_time_tof_2", int_t,  0,  "Integration time TOF2 [uS]",        600, 0, 4000)
gen.add("integration_time_tof_3", int_t,  0,  "Integration time TOF3 [uS]",       4000, 0, 4000)
gen.add("min_amplitude",          int_t,  0,  "threshold minAmplitude LSB",         60, 0, 100000)
gen.add("roi_left_x",          int_t, 0,  "ROI left X",     0, 0,  312)
gen.add("roi_right_x",         int_t, 0,  "ROI right X",  319, 7,  319)
gen.add("roi_height",          int_t, 0,  "ROI height",   240, 8,  240)

exit(gen.generate(PACKAGE, "limu_camera_node", "limu_tof_ros"))
